meadow-client/src/Semantics.purs

```diff
@@ -0,0 +1,1067 @@
+nullifyInvalidUses env (Both contract1 contract2) =
+  Both (nullifyInvalidUses env contract1) (nullifyInvalidUses env contract2)
+nullifyInvalidUses env (Choice observation contract1 contract2) =
+  Choice observation (nullifyInvalidUses env contract1) (nullifyInvalidUses env contract2)
+nullifyInvalidUses env (When observation timeout contract1 contract2) =
+  When observation timeout (nullifyInvalidUses env contract1) (nullifyInvalidUses env contract2)
+nullifyInvalidUses env (While observation timeout contract1 contract2) =
+  While observation timeout (nullifyInvalidUses env contract1) (nullifyInvalidUses env contract2)
+nullifyInvalidUses env (Scale value1 value2 value3 contract) =
+  Scale value1 value2 value3 (nullifyInvalidUses env contract)
+nullifyInvalidUses env (Let letLabel contract1 contract2) =
+  Let letLabel (nullifyInvalidUses env contract1) (nullifyInvalidUses newEnv contract2)
+  where newEnv = addToEnvironment letLabel Null env -- We just need to mark it as available for this function 
+nullifyInvalidUses env (Use letLabel) =
+  case lookupEnvironment letLabel env of
+    Nothing -> Null
+    Just _ -> Use letLabel
+
+-- Replaces a label with another one (unless it is shadowed)
+relabel :: LetLabel -> LetLabel -> Contract -> Contract
+relabel _ _ Null = Null
+relabel startLabel endLabel (Commit idAction idCommit person value timeout1 timeout2 contract1 contract2) =
+  Commit idAction idCommit person value timeout1 timeout2 (relabel startLabel endLabel contract1) (relabel startLabel endLabel contract2)
+relabel startLabel endLabel (Pay idAction idCommit person value timeout contract1 contract2) =
+  Pay idAction idCommit person value timeout (relabel startLabel endLabel contract1) (relabel startLabel endLabel contract2)
+relabel startLabel endLabel (Both contract1 contract2) =
+  Both (relabel startLabel endLabel contract1) (relabel startLabel endLabel contract2)
+relabel startLabel endLabel (Choice observation contract1 contract2) =
+  Choice observation (relabel startLabel endLabel contract1) (relabel startLabel endLabel contract2)
+relabel startLabel endLabel (When observation timeout contract1 contract2) =
+  When observation timeout (relabel startLabel endLabel contract1) (relabel startLabel endLabel contract2)
+relabel startLabel endLabel (While observation timeout contract1 contract2) =
+  While observation timeout (relabel startLabel endLabel contract1) (relabel startLabel endLabel contract2)
+relabel startLabel endLabel (Scale value1 value2 value3 contract) =
+  Scale value1 value2 value3 (relabel startLabel endLabel contract)
+relabel startLabel endLabel (Let letLabel contract1 contract2) =
+  Let letLabel (relabel startLabel endLabel contract1)
+      (if (letLabel == startLabel)
+       then contract2
+       else relabel startLabel endLabel contract2)
+relabel startLabel endLabel (Use letLabel) =
+  if (letLabel == startLabel) then Use endLabel else Use letLabel
```

Comment:
it feels like there's a functor and recursion scheme here that could get rid of all but the last 2 cases, not sure though, just a thought.

Reply:
:+1: